src/Crate/PDO/Http/ServerPool.php

```diff
@@ -0,0 +1,257 @@
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+namespace Crate\PDO\Http;
+
+use Crate\PDO\Exception\RuntimeException;
+use Crate\PDO\PDO;
+use Crate\Stdlib\Collection;
+use Crate\Stdlib\CollectionInterface;
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\BadResponseException;
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\RequestOptions;
+
+final class ServerPool implements ServerPoolInterface
+{
+    private const DEFAULT_SERVER = 'localhost:4200';
+
+    /**
+     * @var string[]
+     */
+    private $availableServers = [];
+
+    /**
+     * @var Client
+     */
+    private $httpClient;
+    /**
+     * @var PDO
+     */
+    private $pdo;
+
+    /**
+     * Client constructor.
+     *
+     * @param PDO   $pdo
+     * @param array $servers
+     */
+    public function __construct(PDO $pdo, array $servers)
+    {
+        if (\count($servers) === 0) {
+            $servers = [self::DEFAULT_SERVER];
+        }
+
+        // micro optimization so we don't always hit the same server first
+        shuffle($servers);
+
+        foreach ($servers as $server) {
+            $this->availableServers[] = $server;
+        }
+
+        $this->pdo        = $pdo;
+        $this->httpClient = new Client();
+    }
+
+    /**
+     * {@Inheritdoc}
+     */
+    public function execute($query, array $parameters): CollectionInterface
+    {
+        while (true) {
+            $nextServer = $this->nextServer();
+
+            $options = $this->getHttpOptions($nextServer, $query, $parameters);
+
+            try {
+                $response     = $this->httpClient->post('/_sql', $options);
+                $responseBody = json_decode($response->getBody(), true);
+
+                return new Collection(
+                    $responseBody['rows'],
+                    $responseBody['cols'],
+                    $responseBody['duration'],
+                    $responseBody['rowcount']
+                );
+
+            } catch (ConnectException $exception) {
+                // drop the server from the list of available servers
+                $this->dropServer($nextServer);
```

Comment:
This can lead to multiple subsequent queries dropping servers. I suggest using aby-red `$visited` parameter, active only for this scope